test/Demangle/Inputs/manglings.txt

```diff
@@ -428,9 +428,8 @@ $s1A3bar1aySSYt_tF ---> A.bar(a: _const Swift.String) -> ()
 $s1t1fyyFSiAA3StrVcs7KeyPathCyADSiGcfu_SiADcfu0_33_556644b740b1b333fecb81e55a7cce98ADSiTf3npk_n ---> function signature specialization <Arg[1] = [Constant Propagated KeyPath : _556644b740b1b333fecb81e55a7cce98<t.Str,Swift.Int>]> of implicit closure #2 (t.Str) -> Swift.Int in implicit closure #1 (Swift.KeyPath<t.Str, Swift.Int>) -> (t.Str) -> Swift.Int in t.f() -> ()
 $s21back_deploy_attribute0A12DeployedFuncyyFTwb ---> back deployment thunk for back_deploy_attribute.backDeployedFunc() -> ()
 $s21back_deploy_attribute0A12DeployedFuncyyFTwB ---> back deployment fallback for back_deploy_attribute.backDeployedFunc() -> ()
-$s4test3FooVAAyyAA1P_pyxqd__XPlF ---> test.Foo.test<A>(test.P<A, A1>) -> ()
-$s4test3FooVAAyyAA1P_pyxXPF ---> test.Foo.test(test.P<A>) -> ()
+$s4test3fooyyAA1P_px1TRtz_XPlF ---> test.foo<A>(any test.P<A.T == A>) -> ()
+$s4test3fooyyAA1P_pSS1TAaCPRtz_Si1UAERtzXPF ---> test.foo(any test.P<A.test.P.T == Swift.String, A.test.P.U == Swift.Int>) -> ()
```

Comment:
Oof, this demangling is super confusing.  I guess we're in a generic context that says τ_0_0 is such-and-such type?  Maybe we can push a different context when the LHS of these?

This ties in with my concern about using requirements here — it's really a false economy, because τ_0_0 means different things in different positions in the requirement.  And you still don't have the ability to express things like `P<.Slice == .self>`, because you can't write τ_0_0 on the RHS without it being ambiguous in case you're actually in a generic context.

Reply:
If you want to stick with mangling requirements, you could just build them in terms of some new "protocol self" base type that's guaranteed to mangle differently from other types.  Or you could use `Requirement` internally in the compiler when building + uniquing things and then do a special-case mangling at the last minute.